lib/srv/db/postgres/proxy.go

```diff
@@ -0,0 +1,163 @@
+/*
+Copyright 2020 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package postgres
+
+import (
+	"context"
+	"crypto/tls"
+	"io"
+	"net"
+	"strings"
+
+	"github.com/gravitational/teleport/lib/auth"
+
+	"github.com/jackc/pgproto3/v2"
+
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+)
+
+// Proxy proxies connections from Postgres clients to database services
+// over reverse tunnel. It runs inside Teleport proxy service.
+//
+// Implements db.DatabaseProxy.
+type Proxy struct {
+	// TLSConfig is the proxy TLS configuration.
+	TLSConfig *tls.Config
+	// Middleware is the auth middleware.
+	Middleware *auth.Middleware
+	// ConnectToSite is used to connect to remote database server over reverse tunnel.
+	ConnectToSite func(context.Context) (net.Conn, error)
+	// Log is used for logging.
+	Log logrus.FieldLogger
+}
+
+// HandleConnection accepts connection from a Postgres client, authenticates
+// it and proxies it to an appropriate database service.
+func (p *Proxy) HandleConnection(ctx context.Context, clientConn net.Conn) (err error) {
```

Comment:
who's responsible for closing `clientConn`?

Reply:
Like mentioned above, it is closed by the dispatcher (db.ProxyServer).